mako/codegen.py

```diff
@@ -785,16 +788,36 @@ def locate_encode(name):
             else:
                 return filters.DEFAULT_ESCAPES.get(name, name)
 
-        if "n" not in args:
+        filter_args = []
+        conflict_n = "%sn" % CONFLICT_PREFIX
+        if conflict_n not in args:
             if is_expression:
                 if self.compiler.pagetag:
                     args = self.compiler.pagetag.filter_args.args + args
-                if self.compiler.default_filters and "n" not in args:
+                    filter_args = self.compiler.pagetag.filter_args.args
+                if self.compiler.default_filters and conflict_n not in args:
                     args = self.compiler.default_filters + args
         for e in args:
-            # if filter given as a function, get just the identifier portion
-            if e == "n":
+            if e == conflict_n:
                 continue
+            if e.startswith(CONFLICT_PREFIX):
```

Comment:
this code is hard to follow because it's entirely separate from the code that was already here, even though it's doing mostly the same thing.    I think the new logic should be integrated with the existing logic rather than using an if: statement around the whole thing.    

If we start by setting up `e` into the two states that are needed:

```
if e.startswith(CONFLICT_PREFIX):
    render_e = e.replace(CONFLICT_PREFIX, "")
    is_default_filter = True
else:
    render_e = e
    is_default_filter = False
```


then the rest of the code can be continued without a big break, using is_default_filter as needed, this lets us see which parts of what we're doing are common to both

I did a step by step refactor, and came up with this, can you review?    

```py
        for e in args:
            if e == conflict_n:
                continue

            if e.startswith(CONFLICT_PREFIX):
                render_e = e.replace(CONFLICT_PREFIX, "")
                is_default_filter = e not in filter_args
            else:
                render_e = e
                is_default_filter = False

            # if filter given as a function, get just the identifier portion
            m = re.match(r"(.+?)(\(.*\))", e)

            if m:
                if not is_default_filter:
                    ident, fargs = m.group(1, 2)
                    f = locate_encode(ident)
                    render_e = f + fargs
                target = "%s(%s)" % (render_e, target)

            elif is_default_filter:
                target = "%s(%s) if %s is not UNDEFINED else %s(%s)" % (
                    render_e,
                    target,
                    render_e,
                    locate_encode(render_e),
                    target,
                )
            else:
                render_e = locate_encode(render_e)
                assert render_e is not None
                target = "%s(%s)" % (render_e, target)

        return target
```